### DISCUSSION TITLE
Should empty `GraphPattern::BGP` short-circuit evaluation?

### DISCUSSION THREAD

**Discussion by hobofan** [oxigraph/oxigraph#321] [Q&A] [answered]
Heyho,

When doing some high-level query optimization, up until now I used an empty `GraphPattern::BGP { patterns: vec![] }` in places where I was able to infer that all matches from that branch of the query would be filtered out anyways, in an attempt to eliminate that branch and avoid potentially expensive query evaluation.
To me, the fact that this is the `GraphPattern` that is returned by the `Default` implementation, also kind of implied that behavior.

However, it looks like this actually triggers this branch in the `plan_builder`, where evaluation continues with an empty tuple, rather than no tuples:
https://github.com/oxigraph/oxigraph/blob/c2fd3920b5de31252bcc96a7217be3663a549600/lib/src/sparql/plan_builder.rs#L79-L81

Naively removing that tuple did cause some test failures, so I'd figure I'd clarify the intended behavior here.

As a workaround, I was able to use `GraphPattern::Values { variables: vec![], bindings: vec![] }` to get the behavior I was looking for.

**Comment by Tpt** [comment, marked as answer]
Hi!

Yes, it's a trap. The evaluation the empty BGP to a single empty tuple is because the evaluation of a BGP is the join of all its elements and the "identity element" of join is a single empty tuple. 

This also works well with the implicit join syntax of SPARQL. If the empty BGP (` ` or `{}` in SPARQL) would return no tuple then the join of nothing and `?s ?p ?o` (`{} ?s ?p ?o.`) would also return no tuple... And that would be just unusable. Hence this interpretation of the empty BGP.

> As a workaround, I was able to use `GraphPattern::Values { variables: vec![], bindings: vec![] }` to get the behavior I was looking for.

Perfect! It sounds like the proper way to do that.
